File: skypydb/embeddings/ollama.py
"""
Ollama embedding functions for vector operations.
"""

import json
import urllib.request
import urllib.error
from typing import List, Optional
from skypydb.embeddings.mixins import EmbeddingCallableMixin, EmbeddingsFunction


class OllamaEmbedding(EmbeddingsFunction, EmbeddingCallableMixin):
    """Ollama embedding function."""

    def __init__(
        self,
        model: str = "mxbai-embed-large",
        base_url: str = "http://localhost:11434",
        dimension: Optional[int] = None,
    ):
        """
        Initialize Ollama embedding function.

        Args:
            model: Name of the Ollama embedding model to use
            base_url: Base URL for Ollama API (default: http://localhost:11434)
        """

        super().__init__(dimension=dimension)
        self.model = model
        self.base_url = base_url.rstrip("/")

    def _get_embedding(
        self,
        text: str,
    ) -> List[float]:
        """
        Get embedding for a single text using Ollama API.

        Args:
            text: Text to embed

        Returns:
            List of floats representing the embedding vector

        Raises:
            ConnectionError: If Ollama server is not reachable
            ValueError: If embedding generation fails
        """

        base_url = getattr(self, "base_url", "http://localhost:11434").rstrip("/")
        model = getattr(self, "model", "mxbai-embed-large")
        url = f"{base_url}/api/embeddings"

        data = json.dumps({
            "model": model,
            "prompt": text
        }).encode("utf-8")

        request = urllib.request.Request(
            url,
            data=data,
            headers={"Content-Type": "application/json"},
            method="POST"
        )

        try:
            with urllib.request.urlopen(request, timeout=60) as response:
                result = json.loads(response.read().decode("utf-8"))
                embedding = result.get("embedding")
                if embedding is None:
                    raise ValueError(
                        f"No embedding returned from Ollama. "
                        f"Make sure model '{model}' is an embedding model."
                    )
                return embedding
        except urllib.error.URLError as e:
            raise ConnectionError(
                f"Cannot connect to Ollama at {base_url}. "
                f"Make sure Ollama is running. If you haven't installed it go to https://ollama.com/download and install it. Error: {e}"
            )
        except json.JSONDecodeError as e:
            raise ValueError(f"Invalid response from Ollama: {e}")

    def get_dimension(self) -> int:
        """
        Get the embedding dimension, generating a test embedding if needed.

        Returns:
            The dimension of embeddings produced by this model
        """

        if self._dimension is None:
            # generate a test embedding to determine dimension
            test_embedding = self._get_embedding("test")
            self._dimension = len(test_embedding)
        return self._dimension

File: skypydb-js/src/embeddings/ollama.ts
import { EmbeddingsFunction } from "./mixins/embeddings_function";
import type { EmbeddingMatrix, EmbeddingVector } from "../types";

type OllamaEmbeddingOptions = {
  model?: string;
  base_url?: string;
  dimension?: number;
};

export class OllamaEmbedding extends EmbeddingsFunction {
  private readonly model: string;
  private readonly base_url: string;

  constructor(options: OllamaEmbeddingOptions = {}) {
    super(options.dimension);
    this.model = options.model ?? "mxbai-embed-large";
    this.base_url = (options.base_url ?? "http://localhost:11434").replace(
      /\/+$/,
      "",
    );
  }

  protected async _get_embedding(text: string): Promise<EmbeddingVector> {
    const response = await fetch(`${this.base_url}/api/embeddings`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        model: this.model,
        prompt: text,
      }),
    });

    if (!response.ok) {
      throw new Error(
        `Cannot connect to Ollama at ${this.base_url}. Status: ${response.status}.`,
      );
    }

    const payload = (await response.json()) as { embedding?: number[] };
    if (!Array.isArray(payload.embedding)) {
      throw new Error(
        `No embedding returned from Ollama. Make sure model '${this.model}' is an embedding model.`,
      );
    }
    return payload.embedding;
  }

  async embed(texts: string[]): Promise<EmbeddingMatrix> {
    if (texts.length === 0) {
      return [];
    }
    return super.embed(texts);
  }
}
